# Day 12: Passage Pathing

## Part one

With your submarine's subterranean subsystems subsisting suboptimally, the only way you're getting out of this cave anytime soon is by finding a path yourself. Not just **a** path - the only way to know if you've found the **best** path is to find **all** of them.

Fortunately, the sensors are still mostly working, and so you build a rough map of the remaining caves (your puzzle input). For example:

```
start-A
start-b
A-c
A-b
b-d
A-end
b-end
```

This is a list of how all of the caves are connected. You start in the cave named `start`, and your destination is the cave named `end`. An entry like b-d means that cave b is connected to cave d - that is, you can move between them.

So, the above cave system looks roughly like this:

```
    start
    /   \
c--A-----b--d
    \   /
     end
```

Your goal is to find the number of distinct **paths** that start at start, end at end, and don't visit small caves more than once. There are two types of caves: **big** caves (written in uppercase, like A) and **small** caves (written in lowercase, like b). It would be a waste of time to visit any small cave more than once, but big caves are large enough that it might be worth visiting them multiple times. So, all paths you find should **visit small caves at most once**, and can **visit big caves any number of times**.

Given these rules, there are `10` paths through this example cave system:

```
start,A,b,A,c,A,end
start,A,b,A,end
start,A,b,end
start,A,c,A,b,A,end
start,A,c,A,b,end
start,A,c,A,end
start,A,end
start,b,A,c,A,end
start,b,A,end
start,b,end
```

(Each line in the above list corresponds to a single path; the caves visited by that path are listed in the order they are visited and separated by commas.)

Note that in this cave system, cave `d` is never visited by any path: to do so, cave b would need to be visited twice (once on the way to cave d and a second time when returning from cave d), and since cave `b` is small, this is not allowed.

Here is a slightly larger example:

```
dc-end
HN-start
start-kj
dc-start
dc-HN
LN-dc
HN-end
kj-sa
kj-HN
kj-dc
```

The `19` paths through it are as follows:

```
start,HN,dc,HN,end
start,HN,dc,HN,kj,HN,end
start,HN,dc,end
start,HN,dc,kj,HN,end
start,HN,end
start,HN,kj,HN,dc,HN,end
start,HN,kj,HN,dc,end
start,HN,kj,HN,end
start,HN,kj,dc,HN,end
start,HN,kj,dc,end
start,dc,HN,end
start,dc,HN,kj,HN,end
start,dc,end
start,dc,kj,HN,end
start,kj,HN,dc,HN,end
start,kj,HN,dc,end
start,kj,HN,end
start,kj,dc,HN,end
start,kj,dc,end
```

Finally, this even larger example has `226` paths through it:

```
fs-end
he-DX
fs-he
start-DX
pj-DX
end-zg
zg-sl
zg-pj
pj-he
RW-he
fs-DX
pj-RW
zg-RW
start-pj
he-WI
zg-he
pj-fs
start-RW
```

**How many paths through this cave system are there that visit small caves at most once?**

In [1]:
def load_input(filepath='./data/input_test.txt'):
    cave_map = {}
    with open('%s' % (filepath, ), 'r') as input_f:
        for input_line in input_f.read().splitlines():
            point_a, point_b = input_line.split('-')
            
            if point_a in cave_map.keys():
                cave_map[point_a].append(point_b)
            else:
                cave_map[point_a] = [point_b]
                
            if point_b in cave_map.keys():
                cave_map[point_b].append(point_a)
            else:
                cave_map[point_b] = [point_a]
    
    return cave_map

cave_map = load_input('./data/input.txt')
cave_map

{'yb': ['pi', 'KN', 'UF', 'xd', 'qx', 'jg'],
 'pi': ['yb'],
 'jg': ['ej', 'KN', 'LD', 'yb', 'UF', 'qx'],
 'ej': ['jg', 'start', 'qx', 'LD'],
 'KN': ['yb', 'jg', 'end'],
 'LD': ['start', 'qx', 'jg', 'xd', 'ej'],
 'start': ['LD', 'qx', 'ej'],
 'end': ['UF', 'xd', 'KN'],
 'UF': ['end', 'yb', 'qx', 'jg', 'xd'],
 'xd': ['yb', 'end', 'LD', 'DM', 'UF'],
 'qx': ['yb', 'start', 'LD', 'ej', 'UF', 'jg'],
 'DM': ['xd']}

In [2]:
def advance_one_step(paths, cave_map):
    new_paths = []
    for path in paths:
        last_point = path[-1]
        
        if last_point == 'end':
            new_paths.append(path.copy())
            continue
        
        for possible_next_points in cave_map[last_point]:
            if possible_next_points == 'start' or (possible_next_points.islower() and possible_next_points in path):
                continue
            else:
                updated_path = path.copy()
                updated_path.append(possible_next_points)
                new_paths.append(updated_path)
                
    return new_paths

def all_paths_found(paths):
    for path in paths:
        if path[-1] != 'end':
            return False
    return True

cave_map = load_input()

paths = [['start']]
while not all_paths_found(paths):
    paths = advance_one_step(paths, cave_map)

print('every paths found: %s' % (all_paths_found(paths), ))
print('number of paths: %s' % (len(paths), ))
paths

every paths found: True
number of paths: 10


[['start', 'A', 'c', 'A', 'b', 'A', 'end'],
 ['start', 'A', 'c', 'A', 'b', 'end'],
 ['start', 'A', 'c', 'A', 'end'],
 ['start', 'A', 'b', 'A', 'c', 'A', 'end'],
 ['start', 'A', 'b', 'A', 'end'],
 ['start', 'A', 'b', 'end'],
 ['start', 'A', 'end'],
 ['start', 'b', 'A', 'c', 'A', 'end'],
 ['start', 'b', 'A', 'end'],
 ['start', 'b', 'end']]

In [3]:
cave_map = load_input('./data/input.txt')

paths = [['start']]
while not all_paths_found(paths):
    paths = advance_one_step(paths, cave_map)

print('every paths found: %s' % (all_paths_found(paths), ))
print('number of paths: %s' % (len(paths), ))

every paths found: True
number of paths: 4011


Your puzzle answer was `4011`.

## Part two

After reviewing the available paths, you realize you might have time to visit a single small cave **twice**. Specifically, big caves can be visited any number of times, a single small cave can be visited at most twice, and the remaining small caves can be visited at most once. However, the caves named start and end can only be visited **exactly once each**: once you leave the start cave, you may not return to it, and once you reach the end cave, the path must end immediately.

Now, the 36 possible paths through the first example above are:

```
start,A,b,A,b,A,c,A,end
start,A,b,A,b,A,end
start,A,b,A,b,end
start,A,b,A,c,A,b,A,end
start,A,b,A,c,A,b,end
start,A,b,A,c,A,c,A,end
start,A,b,A,c,A,end
start,A,b,A,end
start,A,b,d,b,A,c,A,end
start,A,b,d,b,A,end
start,A,b,d,b,end
start,A,b,end
start,A,c,A,b,A,b,A,end
start,A,c,A,b,A,b,end
start,A,c,A,b,A,c,A,end
start,A,c,A,b,A,end
start,A,c,A,b,d,b,A,end
start,A,c,A,b,d,b,end
start,A,c,A,b,end
start,A,c,A,c,A,b,A,end
start,A,c,A,c,A,b,end
start,A,c,A,c,A,end
start,A,c,A,end
start,A,end
start,b,A,b,A,c,A,end
start,b,A,b,A,end
start,b,A,b,end
start,b,A,c,A,b,A,end
start,b,A,c,A,b,end
start,b,A,c,A,c,A,end
start,b,A,c,A,end
start,b,A,end
start,b,d,b,A,c,A,end
start,b,d,b,A,end
start,b,d,b,end
start,b,end
```

The slightly larger example above now has `103` paths through it, and the even larger example now has 3509 paths through it.

Given these new rules, **how many paths through this cave system are there?**

In [4]:
def advance_one_step_ex2(paths, cave_map):
    new_paths = []
    for path in paths:
        # print('path: %s' % (path, ))
        last_point = path['path'][-1]
        has_visited_s_cave_twice = path['has_visited_s_cave_twice']
        # print('last_point: %s' % (last_point, ))
        
        if last_point == 'end':
            new_paths.append(path.copy())
            continue
        
        for possible_next_points in cave_map[last_point]:
            # print('possible_next_points: %s' % (possible_next_points, ))
            if possible_next_points == 'start' or (possible_next_points.islower() and has_visited_s_cave_twice and possible_next_points in path['path']):
                # print('possible_next_points: %s' % (possible_next_points, ))
                continue
            else:
                updated_path = {}
                updated_path['path'] = path['path'].copy()
                updated_path['path'].append(possible_next_points)
                if has_visited_s_cave_twice or (possible_next_points.islower() and possible_next_points in path['path']):
                    updated_path['has_visited_s_cave_twice'] = True
                else:
                    updated_path['has_visited_s_cave_twice'] = False
                new_paths.append(updated_path)
                
    return new_paths

def all_paths_found_ex2(paths):
    for path in paths:
        if path['path'][-1] != 'end':
            return False
    return True

cave_map = load_input()

paths = [{'path': ['start'], 'has_visited_s_cave_twice': False}]
while not all_paths_found_ex2(paths):
    paths = advance_one_step_ex2(paths, cave_map)

print('every paths found: %s' % (all_paths_found_ex2(paths), ))
print('number of paths: %s' % (len(paths), ))
paths

every paths found: True
number of paths: 36


[{'path': ['start', 'A', 'c', 'A', 'c', 'A', 'b', 'A', 'end'],
  'has_visited_s_cave_twice': True},
 {'path': ['start', 'A', 'c', 'A', 'c', 'A', 'b', 'end'],
  'has_visited_s_cave_twice': True},
 {'path': ['start', 'A', 'c', 'A', 'c', 'A', 'end'],
  'has_visited_s_cave_twice': True},
 {'path': ['start', 'A', 'c', 'A', 'b', 'A', 'c', 'A', 'end'],
  'has_visited_s_cave_twice': True},
 {'path': ['start', 'A', 'c', 'A', 'b', 'A', 'b', 'A', 'end'],
  'has_visited_s_cave_twice': True},
 {'path': ['start', 'A', 'c', 'A', 'b', 'A', 'b', 'end'],
  'has_visited_s_cave_twice': True},
 {'path': ['start', 'A', 'c', 'A', 'b', 'A', 'end'],
  'has_visited_s_cave_twice': False},
 {'path': ['start', 'A', 'c', 'A', 'b', 'd', 'b', 'A', 'end'],
  'has_visited_s_cave_twice': True},
 {'path': ['start', 'A', 'c', 'A', 'b', 'd', 'b', 'end'],
  'has_visited_s_cave_twice': True},
 {'path': ['start', 'A', 'c', 'A', 'b', 'end'],
  'has_visited_s_cave_twice': False},
 {'path': ['start', 'A', 'c', 'A', 'end'], 'has_

In [5]:
cave_map = load_input('./data/input.txt')

paths = [{'path': ['start'], 'has_visited_s_cave_twice': False}]
while not all_paths_found_ex2(paths):
    paths = advance_one_step_ex2(paths, cave_map)

print('every paths found: %s' % (all_paths_found_ex2(paths), ))
print('number of paths: %s' % (len(paths), ))

every paths found: True
number of paths: 108035


Your puzzle answer was `108035`.